### ISSUE TITLE
Timeout issue workaround!

### ISSUE THREAD

**Issue by pachacamac** [patriksimek/vm2#355]
I found a way that helped me get around the timeout limitation for NodeVM and thought I'd share it in case someone has the same issue. What I ended up doing is to run the NodeVM in a separate child process. This allows me to safely kill the child process when the timeout is reached since the parent process is not affected by the `while(true);` issue and is still able to kill the child. :crossed_swords: :drop_of_blood: :skull_and_crossbones: 

Now before you get too excited there is one downside to this method! **The communication between the parent process and the child process is limited to JSON**. That means you can not push functions and the like from the parent to the child or vise versa. But if this is not an issue for you this might just be the solution you were looking for :smile: 

Example implementation:

```javascript
const {NodeVM} = require('vm2');
const cp = require('child_process');

// this function can safely be used from the main process. sandbox is limited to JSON values
function run(code='', sandbox={}, timeout=5000){
  return new Promise((resolve, reject)=>{
    const sub = cp.fork(__filename); // fork same file as child process. see code at the end
    const kto = setTimeout(()=>{sub.kill(9)}, timeout);
    sub.on('message', (m)=>{
      if(m.type == 'finished'){
        clearTimeout(kto);
        resolve({returnValue: m.returnValue, sandbox: m.sandbox});
      }
    });
    sub.on('exit', function(c,s) {
      if(c !== 0) reject(new Error(`Execution halted! ${s}`));
    });
    sub.send({code, sandbox});
  })
}

// this function is only used from the spawned child process
async function runVM(code='', sandbox={}){
  code = `module.exports = async function(){ ${code} }`;
  sandbox = Object.assign({wait: (n)=>new Promise(r=>setTimeout(r, n))}, sandbox); // you can still push functions to the sandbox if they're defined in the child process!
  const vm = new NodeVM({wasm: false, eval: true, sandbox});
  const resultfn = vm.run(code);
  const returnValue = await resultfn();
  return {sandbox, returnValue};
}

module.exports = { run }

// ran as script and not imported. used as child process
if(typeof require !== 'undefined' && require.main === module){
  (async function(){
    process.on('message', async function(m) {
      if(m.code && m.sandbox){
        const result = await runVM(m.code, m.sandbox);
        process.send({type: 'finished', returnValue: result.returnValue, sandbox: result.sandbox});
        process.exit();
      }
    });
  })();
}
```

Please let me know if this was helpful and if you find any potential problems or you got other feedback!

**Comment by wintertime-inc** [comment]
I like this idea, but currently thinking of having similar effect while keeping flexibility of injecting objects and functions.

**Comment by pachacamac** [comment]
Awesome, and how would you go about it currently? What's your idea?

**Comment by wintertime-inc** [comment]
https://medium.com/@bvjebin/js-infinite-loops-killing-em-e1c2f5f2db7f

**Comment by pachacamac** [comment]
But that would require that you disable eval support right? Otherwise couldn't you just generate code at runtime that creates an infinite loop again?

**Comment by edopalmieri** [comment]
Hi, 
I have used [Node's worker threads](https://nodejs.org/api/worker_threads.html) to implement a similar solution.
I believe comms between threads when using worker threads is a bit more advanced than with child processes. Although I cant comment too much as I am only using it to exchange objects at the moment. 

I hope it can help.
Cheers

**Comment by pachacamac** [comment]
Thanks @edopalmieri I'll check out node worker threads. Quick glance at the [relevant docs here](https://nodejs.org/api/worker_threads.html#worker_threads_port_postmessage_value_transferlist) and [here](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) seem to allow definitely more than what JSON has to offer however unfortunately it doesn't seem like functions are part of it. I'll play with it a bit and take it for a test drive :)

**Comment by y21** [comment]
`worker_threads` has the advantage that it doesn't require spawning a new process and creates a new V8 isolate for the worker in the same process which lets you transfer objects between V8 contexts more easily, however this approach also has a (few) downside/s. The timeout option relies on V8 being able to terminate execution and it works well for simple infinite loops like `for(;;);`, however I've noticed it doesn't work very well when in the middle of executing native code. One example being the BigInt constructor. IIUC, this is now fixed, but the BigInt constructor used to not have any interrupt checks and something like `BigInt("1".repeat(1e8))` would block the thread for a very long time. Of course this is just one example, and there might be other cases where execution can't be terminated. This one takes about 20-40 seconds to terminate on my machine:
```js
const { isMainThread, Worker } = require("worker_threads");

if (isMainThread) {
    const w = new Worker(__filename);
    let createdAt;
    w.on("online", () => {
        // Give worker some time to get to vm.run
        setTimeout(() => {
            createdAt = Date.now();
            w.terminate();
        }, 1000);
    });
    w.on("exit", () => {
        console.log("Took %dms to terminate worker", Date.now() - createdAt);
    });
} else {
    const { VM } = require("vm2");
    const vm = new VM({ timeout: 100, eval: false });
    try {
        // Dangerous, untrusted code
        vm.run(`~~~~~encodeURI('€'.repeat(0x1fffffe8))`);
    } catch (e) {
        console.log(e);
    }
}
```
Workers do get their own thread so it's not a huge problem in that it doesn't affect the main thread if a long blocking function in another thread can't be terminated, but a core is still being kept busy so it's not very ideal.

Another problem is V8 OOM errors. AFAIK, V8 out of memory errors are unrecoverable and the process must abort if it happens. Unfortunately it is very easy to trigger this. One can create a very big array or a string. This can bring down the whole process, even if caused in vm2 *in* a worker thread:
```js
const { isMainThread, Worker } = require("worker_threads");

if (isMainThread) {
    const w = new Worker(__filename);
} else {
    const { VM } = require("vm2");
    const vm = new VM({ timeout: 100, eval: false });
    vm.run(`'1'.repeat(0x7ffffff).split('');`);
}
```

IMO it's best to have untrusted code run on its own process because there are always new fun ways to crash V8, or at least have critical parts of your application run somewhere else.
